tests/spread/plugins/craft-parts/build-and-run-hello/matter-hello/app/main.cpp

```diff
@@ -0,0 +1,103 @@
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "LightingAppCommandDelegate.h"
+#include "LightingManager.h"
+#include <AppMain.h>
+
+#include <app-common/zap-generated/ids/Attributes.h>
+#include <app-common/zap-generated/ids/Clusters.h>
+#include <app/ConcreteAttributePath.h>
+#include <app/server/Server.h>
+#include <lib/support/logging/CHIPLogging.h>
+
+#if defined(CHIP_IMGUI_ENABLED) && CHIP_IMGUI_ENABLED
+#include <imgui_ui/ui.h>
+#include <imgui_ui/windows/light.h>
+#include <imgui_ui/windows/occupancy_sensing.h>
+#include <imgui_ui/windows/qrcode.h>
+
+#include <iostream>
+
+#endif
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::app::Clusters;
+
+namespace {
+
+constexpr char kChipEventFifoPathPrefix[] = "/tmp/chip_lighting_fifo_";
+NamedPipeCommands sChipNamedPipeCommands;
+LightingAppCommandDelegate sLightingAppCommandDelegate;
+} // namespace
+
+void MatterPostAttributeChangeCallback(const chip::app::ConcreteAttributePath & attributePath, uint8_t type, uint16_t size,
+                                       uint8_t * value)
+{
+    if (attributePath.mClusterId == OnOff::Id && attributePath.mAttributeId == OnOff::Attributes::OnOff::Id)
+    {
+        LightingMgr().InitiateAction(*value ? LightingManager::ON_ACTION : LightingManager::OFF_ACTION);
+    }
+}
+
+/** @brief OnOff Cluster Init
+ *
+ * This function is called when a specific cluster is initialized. It gives the
+ * application an opportunity to take care of cluster initialization procedures.
+ * It is called exactly once for each endpoint where cluster is present.
+ *
+ * @param endpoint   Ver.: always
+ *
+ * TODO Issue #3841
+ * emberAfOnOffClusterInitCallback happens before the stack initialize the cluster
+ * attributes to the default value.
+ * The logic here expects something similar to the deprecated Plugins callback
+ * emberAfPluginOnOffClusterServerPostInitCallback.
+ *
+ */
+void emberAfOnOffClusterInitCallback(EndpointId endpoint)
+{
+    // TODO: implement any additional Cluster Server init actions
+}
+
+void ApplicationInit()
+{
+    std::string path = kChipEventFifoPathPrefix + std::to_string(getpid());
+
+    if (sChipNamedPipeCommands.Start(path, &sLightingAppCommandDelegate) != CHIP_NO_ERROR)
+    {
+        ChipLogError(NotSpecified, "Failed to start CHIP NamedPipeCommands");
+        sChipNamedPipeCommands.Stop();
+    }
+}
+
+void ApplicationShutdown()
+{
+    if (sChipNamedPipeCommands.Stop() != CHIP_NO_ERROR)
+    {
+        ChipLogError(NotSpecified, "Failed to stop CHIP NamedPipeCommands");
+    }
+}
+
+int main(int argc, char * argv[])
+{
+    std::cout << "hello world" << std::endl;
+
+    return 0;
+}
```

Comment:
This isn't performing the [app initializations](https://github.com/project-chip/connectedhomeip/blob/a4300e7f3a8999e5b283aeae20e39d645b296a67/examples/lighting-app/linux/main.cpp#L98-L116). Does the test assertion expect "hello world" and nothing else?

Reply:
The app initialization part has been removed because, based on this file, it seems like the test section only expects 'hello world': https://github.com/snapcore/snapcraft/blob/main/tests/spread/plugins/craft-parts/build-and-run-hello/task.yaml#L56

Reply:
That is with the hello word assertion. It seems relatively easy to create [other tests](https://github.com/snapcore/snapcraft/tree/ffbdabf53de21606d9d12e15f16d4bca5e4423a5/tests/spread/plugins/craft-parts).

The application builds so we could assume that the included libraries are present in shared object files:
```
:: [597/597] stamp obj/default.stamp                                                                 
:: + ldd out/build/chip-lighting-app                                                                 
::   linux-vdso.so.1 (0x00007fffd252a000)                                                            
::   libcrypto.so.3 => /lib/x86_64-linux-gnu/libcrypto.so.3 (0x00007fc4b5c00000)                     
::   libgio-2.0.so.0 => /lib/x86_64-linux-gnu/libgio-2.0.so.0 (0x00007fc4b5a27000)                   
::   libgobject-2.0.so.0 => /lib/x86_64-linux-gnu/libgobject-2.0.so.0 (0x00007fc4b61a0000)           
::   libglib-2.0.so.0 => /lib/x86_64-linux-gnu/libglib-2.0.so.0 (0x00007fc4b6066000)                 
::   libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007fc4b5600000)                     
::   libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fc4b5940000)                               
::   libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fc4b5200000)                               
::   /lib64/ld-linux-x86-64.so.2 (0x00007fc4b64a8000)                                                
::   libgmodule-2.0.so.0 => /lib/x86_64-linux-gnu/libgmodule-2.0.so.0 (0x00007fc4b605f000)           
::   libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007fc4b6043000)                               
::   libmount.so.1 => /lib/x86_64-linux-gnu/libmount.so.1 (0x00007fc4b58fc000)                       
::   libselinux.so.1 => /lib/x86_64-linux-gnu/libselinux.so.1 (0x00007fc4b58d0000)                   
::   libffi.so.8 => /lib/x86_64-linux-gnu/libffi.so.8 (0x00007fc4b58c3000)                           
::   libpcre.so.3 => /lib/x86_64-linux-gnu/libpcre.so.3 (0x00007fc4b584d000)                         
::   libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007fc4b582d000)                       
::   libblkid.so.1 => /lib/x86_64-linux-gnu/libblkid.so.1 (0x00007fc4b55c9000)                       
::   libpcre2-8.so.0 => /lib/x86_64-linux-gnu/libpcre2-8.so.0 (0x00007fc4b5532000)                   
:: + mkdir -p /root/parts/lighting-hello/install/bin                                                 
:: + cp out/build/chip-lighting-app /root/parts/lighting-hello/install/bin/lighting-app              
Staging lighting-hello 
```

One thing that the plugin does but isn't tested is the storage path replacement. 

If you intend to follow the approach to partially test, I don't see any reason for replicating the whole main file with mostly unused code. The print and return can be simply injected into the example code.

Reply:
> The snapcraft team requires the unit tests and a spread test (hello world snap). Of cource that the automated testing of storage path replacement can be performed. Do you know where is the best place we can place this test?

The hello world snap doesn't have to print just "Hello World" and exit. The "Hello World" printing is just for one of the readily available spread tests. Other examples are available [here](https://github.com/snapcore/snapcraft/tree/ffbdabf53de21606d9d12e15f16d4bca5e4423a5/tests/spread/plugins/craft-parts), as linked in my previous comment and [mentioned](https://github.com/snapcore/snapcraft/pull/4491#pullrequestreview-1784399034) by the snapcraft team.

Reply:
> The print and return can be simply injected into the example code.

Can you think of a good way to inject code into `int mian()`? I tried using regex, but it can't be done easily due to various numbers of brackets.

Reply:
It should be possible to insert after the match with `sed`. But this isn't needed anymore since you've decided to create custom tests. In those tests, you can check for the SDK initialization message without having to modify the reference example code.